tests/Highlighter.py

```diff
@@ -0,0 +1,358 @@
+
+                else:
+                    if type(rule["token"]) != list :
+                        raise ValueError(f"if regex has groups, token must be a list. In this rule : {rule}!")
+
+                    else:
+                        if rule["nb_groups"] != len(rule["token"]):
+                            raise ValueError(f"The number of groups in the regex is different from the number of tokens. In this rule : {rule}!")
+
+    # apply rules on code
+    def highlight(self, code):
+        outputs = []
+        token = "start"
+        for line in code.split("\n"):
+            output, token = self.highlight_rules_line(line, token)
+            outputs.append(output)
+        return "\n".join(outputs)
+
+    # apply rule in one line of code
+    def highlight_rules_line(self, code, start_token="start"):
+        # Initialisation output
+        output = []
+
+        # Initialisation variables
+        current_state = start_token
+        current_position = 0
+
+        default_token = "text"
+
+        end_of_search = False
+        while not end_of_search:
+
+            find_transition, next_transition = self.find_match(code, current_state, current_position)
+
+            if find_transition :
+                # Application of coloring on the code
+
+                # get match
+                current_rule, current_match = next_transition["rule"], next_transition["match"]
+
+                # we color the characters since the last match with the default coloring
+                for c in range(current_position, current_match.start()):
+                    output.append(TOKEN_CODE[default_token])
+
+                # we recover the next default coloring
+                if "defaultToken" in current_rule:
+                    default_token = current_rule["defaultToken"]
+                else:
+                    default_token = "text"
+
+        
+                # if rule has only one groups
+                if current_rule["nb_groups"] == 0:
+                    tok    = current_rule['token']
+                    start  = current_match.start()
+                    length = current_match.end() - current_match.start()
+                    output.append(TOKEN_CODE[tok] * length)
+
+                # if rule has only multiples groups
+                else:
+                    pos = current_match.start()
+                    for i, submatch in enumerate(current_match.groups()):    
+                        tok = current_rule['token'][i%len(current_rule['token'])]        
+                        output.append(TOKEN_CODE[tok] * len(submatch))
+                        pos += len(submatch)
+
+                current_position = current_match.end()
+
+                if 'next' in current_rule:
+                    current_state = current_rule['next']
+
+
+                if current_position == len(code):
+                    end_of_search = True
+
+            else:
+                end_of_search = True
```

Comment:
Personally I'm not a fan of very long `if` blocks:

```py
if success_condition:
  # a lot of code
  #
  # 
  # no but really, a lot of lines here
  #
  # yep this is happening
  # 
  # scrolling is definitely required
else:
  # short code to deal with failure case
```

Because they hurt readability (it's hard to find the `else` clause from that `if`). I generally write in an early-exit style, and I recommend you get used to doing the same:

```py
if not success_condition: 
  # deal with error case
  return; break; continue; # whatever is necessary to stop execution

# this is where the long happy path starts
#
# can still be long but it's obvious
```

Reply:
In this case I might even get rid of the boolean flag altogether and do something like this:

```py
next_transition = self.find_match(code, current_state, current_position)
while next_transition:
  start, end = transition_position(next_transition)
  mark_as_text(current_position, start)
  mark_transition(next_transition)
  current_position = end

  next_transition = self.find_match(code, current_state, current_position)

mark_as_text(current_position, len(code))
```

Reply:
Not trying to force you to change anything, just giving you something to think about.